server/src/main/java/org/opensearch/action/search/TransportDeletePitAction.java

```diff
@@ -0,0 +1,122 @@
+ */
+public class TransportDeletePitAction extends HandledTransportAction<DeletePitRequest, DeletePitResponse> {
+    private final NamedWriteableRegistry namedWriteableRegistry;
+    private TransportSearchAction transportSearchAction;
+    private final ClusterService clusterService;
+    private final SearchTransportService searchTransportService;
+    private static final Logger logger = LogManager.getLogger(TransportDeletePitAction.class);
+
+    @Inject
+    public TransportDeletePitAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        NamedWriteableRegistry namedWriteableRegistry,
+        TransportSearchAction transportSearchAction,
+        ClusterService clusterService,
+        SearchTransportService searchTransportService
+    ) {
+        super(DeletePitAction.NAME, transportService, actionFilters, DeletePitRequest::new);
+        this.namedWriteableRegistry = namedWriteableRegistry;
+        this.transportSearchAction = transportSearchAction;
+        this.clusterService = clusterService;
+        this.searchTransportService = searchTransportService;
+    }
+
+    /**
+     * Invoke 'delete all pits' or 'delete list of pits' workflow based on request
+     */
+    @Override
+    protected void doExecute(Task task, DeletePitRequest request, ActionListener<DeletePitResponse> listener) {
+        List<String> pitIds = request.getPitIds();
+        if (pitIds.size() == 1 && "_all".equals(pitIds.get(0))) {
+            deleteAllPits(listener);
+        } else {
+            deletePits(listener, request);
+        }
+    }
+
+    /**
+     * Deletes list of pits, return success if all reader contexts are deleted ( or not found ).
+     */
+    private void deletePits(ActionListener<DeletePitResponse> listener, DeletePitRequest request) {
+        List<SearchContextIdForNode> contexts = new ArrayList<>();
+        for (String pitId : request.getPitIds()) {
+            SearchContextId contextId = SearchContextId.decode(namedWriteableRegistry, pitId);
+            contexts.addAll(contextId.shards().values());
+        }
+        ActionListener<Integer> deleteListener = ActionListener.wrap(r -> {
+            if (r == contexts.size()) {
+                listener.onResponse(new DeletePitResponse(true));
+            } else {
+                logger.debug(() -> new ParameterizedMessage("Delete PITs failed. " + "Cleared {} contexts out of {}", r, contexts.size()));
+                listener.onResponse(new DeletePitResponse(false));
+            }
+        }, e -> {
+            logger.debug("Delete PITs failed ", e);
+            listener.onResponse(new DeletePitResponse(false));
+        });
+        SearchUtils.deletePits(contexts, deleteListener, clusterService.state(), searchTransportService);
+    }
+
+    /**
+     * Delete all active PIT reader contexts
+     */
+    private void deleteAllPits(ActionListener<DeletePitResponse> listener) {
```

Comment:
this logic is wrong. One PIT Id can be created across clusters. Hence here we should do a list all PITs and then perform delete list of PITs so that the PITs holding contexts in other clusters can also be deleted.